ocaml-lsp-server/src/doc_to_md.ml

```diff
@@ -165,6 +212,68 @@ let rec nestable_block_element_to_block
     let meta = loc_to_meta location in
     Block.Ext_math_block (code_block, meta)
 
+and nestable_block_element_to_inlines
```

Comment:
This function might be a lot of maintenance and hurts the content.
Would it be reasonable to use raw HTML tags for tables ?

Reply:
I agree that this function is not ideal...

An advantage of not using proper table compared to raw html is that the output might be more readable even without translating it to html. Also, we get cmarkit type guaranty.

Raw html tags could be an idea, but their start-end semantics not that simple (https://spec.commonmark.org/0.30/#html-blocks).

In particular, it seems to break out of the containing block. For instance,

```
- first item
  + first subitem
  + second subitem
```
renders by github as:

- first item
  + first subitem
  + second subitem

while with html inside (the blank line is the end condition fot the uninterpreted html)

```
- first item
  + first subitem
<table><tr><td>

useless table
</td></tr></table>

  + second subitem
```
renders as

- first item
  + first subitem
<table><tr><td>

useless table
</td></tr></table>

  + second subitem

@voodoos would you like me to investigate the "raw html tag" solution?

Reply:
> In particular, it seems to break out of the containing block

So in the case of tables the content would break out of the table ?

Reply:
> So in the case of tables the content would break out of the table ?

What I meant is that the _whole table_ would break out of its _container block_ (recall that all of this is because, in odoc, tables are allowed to be nested in blocks, while it is not allowed in markdown).

For instance, in the example above, I had in mind a table inside a list (which is possible in odoc).

Julow suggested (IIUC) to render in this case the table, not by using markdown tables, but by embedding raw `<table>`/`<tr>`/`<td>` before the table/line/cell, and the corresponding closing tag after.
The markdown example above was thus the translation of a (nested) list containing a table, using this method.

However, we can see, from how github renders it, that in fact the table is not interpreted as inside the list, but "toplevel", and thus also breaks the rest of the list.
(I included a nested list to make it more visible that the included html breaks the flow of the list interpretation, but it might have created noise!)

It seems that cmarkit type system [won't enforce constraint](https://github.com/dbuenzli/cmarkit/blob/main/src/cmarkit.mli#L802) on raw html embedding, which is frightening!

Reply:
So that's not suitable since it would also break more visibly things like numbered lists (see below).
So far I think the more convincing workaround is your current proposition. 

```
- first item
  1. first subitem
<table><tr><td>

useless table
</td></tr></table>

  2. second subitem
  ```

- first item
  1. first subitem
<table><tr><td>

useless table
</td></tr></table>

  2. second subitem

Reply:
Update: The example was wrong, it is in fact possible to have html inside list items in commonmark.

(With the help of cmarkit) I could generate the right markdown for the example above:

So:
```
1. first item
   1. first subitem
      <table><tr><td>
      
      useless table
      </td></tr></table>
      
   2. second subitem
```
renders as

1. first item
   1. first subitem
      <table><tr><td>
      
      useless **table**
      </td></tr></table>
      
   2. second subitem

So I think the suggestion to use html is still valid, and the discussion open again. Speaking of tables, let me use one to summarize the differences:

| | Current approach | Html embedding |
|-|-|-|
|Source| :orange_circle: Source is not very readable due to tables not properly formatted, and blocks turned into inlines | :orange_circle: Source is not very readable, due to html tags and blank lines added all over the place |
|Faithful rendering| :red_circle: The content is modified | :green_circle: The content is preserved |
|Compatibility| :orange_circle: The renderer needs to know about tables, which is not in the commonmark standard | :orange_circle: The parser needs to follow the commonmark standard, which is different from the original one if I understand correctly |
|Rendering target | :green_circle: No constraint | :red_circle: Has to be HTML/markdown, otherwise (eg in latex) embedded html will certainly be ignored |
|Scariness| :green_circle: Not scary, a problem won't leak outside of the table | :orange_circle: Html embedding is a litlle bit scary as making a mistake can break the whole output...|

If the "rendering target" can handle embedded html, maybe the html embedding approach is better? The "content is preserved" is quite a strong benefit, I guess...

Reply:
According to [the specification](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#markupContent), LSP clients that have the capability to parse markdown should follow the Github Flavored Markdown Specification.

That plays in favor of Html but they also warn that:
>  *Please Note* that clients might sanitize the return markdown. A client could
>   decide to remove HTML from the markdown to avoid script execution.

Reply:
And there is more: the client can advertise which tags will actually be allowed in markdown with the "[MarkdownClientCapabilities](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#markdownClientCapabilities).allowedTags" capability. 

It's clear that if we want to respect the protocol we need to have at least two fallbacks:
- A text-only version when markdown is not allowed. (Right now `ocaml-lsp` simply prints the raw comment which is okay)
- A markdown without html when required tags are not allowed

And we could do better by using raw html when the client accepts it.

In any case we need to keep the current version, and I'm not sure having the additional html one is really worth it ?

Reply:
> In any case we need to keep the current version, and I'm not sure having the additional html one is really worth it ?

I agree with this analysis, and I do not think it is worth it: in the current approach, the content is only hurt in the case of table/lists inside tables, which I think will happen very seldomly.

If it turns out a problem, we can add the html approach later.